### DISCUSSION TITLE
Hotkeys in GlotPress and in GlotDict

### DISCUSSION THREAD

**Discussion by amieiro** [GlotPress/GlotPress-WP#1415] [Ideas]
I have been reviewing a small improvement, but that could have an interesting impact in the GlotPress and in the [translate.wordpress.org](https://translate.wordpress.org/) users: improving the GlotPress hotkeys.

## GlotPress

We can find some documentation in the GlotPress wiki about the [Keyboard shortcuts](https://github.com/GlotPress/GlotPress/wiki/1.4-Keyboard-shortcuts) (not updated). The code (started [10 years ago](https://github.com/GlotPress/GlotPress/blame/dd231e278ad30bc97adea029a768555b586a712f/assets/js/editor.js#L138)) can [be found here](https://github.com/GlotPress/GlotPress/blob/develop/assets/js/editor.js#L138-L194). It has these hotkeys:

<figure class="wp-block-table"><table class="has-fixed-layout"><thead><tr><th><strong>Action</strong></th><th><strong>Hotkey</strong></th></tr></thead><tbody><tr><td>Cancel (Hide the editor)</td><td>Escape or Ctrl-Shift-Z</td></tr><tr><td>Previous editor</td><td>Page Up or Ctrl-Up Arrow</td></tr><tr><td>Next editor</td><td>Page Down or Ctrl-Down Arrow</td></tr><tr><td>Save</td><td>Shift-Enter</td></tr><tr><td>Copy original</td><td>Ctrl-Enter or Ctrl-Shift-B</td></tr><tr><td>Approve</td><td>Ctrl-+ or Ctrl-Shift-A</td></tr><tr><td>Reject</td><td>Ctrl-- or Ctrl-Shift-R</td></tr><tr><td>Fuzzy</td><td>Ctrl-~ or Ctrl-Shift-~</td></tr></tbody></table></figure>

## GlotDict

[GlotDict](https://github.com/Mte90/GlotDict) is a browser extension who implements many new features to improve the quality of translation and speed up the process on [translate.wordpress.org](https://translate.wordpress.org/). It is not focused on GlotPress. One of these features is the [hotkeys](https://github.com/Mte90/GlotDict#hotkeys). Currently, it has these hotkeys:

<figure class="wp-block-table"><table class="has-fixed-layout"><thead><tr><th>Action</th><th>Hotkey</th></tr></thead><tbody><tr><td>Add Glossary definition in the translation area</td><td>Right click on a Glossary term</td></tr><tr><td>Add non-breaking spaces near symbols</td><td>Ctrl+Shift+X</td></tr><tr><td>Approve</td><td>Ctrl+Shift+A</td></tr><tr><td>Cancel</td><td>Ctrl+Shift+Z</td></tr><tr><td>Copy from original</td><td>Ctrl+Shift+B</td></tr><tr><td>Dismiss validation warnings for the current visible editor</td><td>Ctrl+D</td></tr><tr><td>Dismiss all validation warnings</td><td>Ctrl+Shift+D</td></tr><tr><td>Force suggest or Force save translation</td><td>Ctrl+Shift+Enter</td></tr><tr><td>Fuzzy</td><td>Ctrl+Shift+F</td></tr><tr><td>Load all consistency suggestions</td><td>Alt+C</td></tr><tr><td>Open next string editor</td><td>Page Down</td></tr><tr><td>Open previous string editor</td><td>Page Up</td></tr><tr><td>Reject</td><td>Ctrl+Shift+R</td></tr><tr><td>Suggest or Save translation</td><td>Ctrl+Enter</td></tr></tbody></table></figure>

## Analysis

<p>We have some duplicated actions with the same hotkeys. E.g.:</p>

<ul><li>Approve:<ul><li>GlotPress: Ctrl-+ or <strong>Ctrl-Shift-A</strong>.</li></ul></li><li>GlotDict: <strong>Ctrl+Shift+A</strong></li></ul>

<p>We have some duplicated actions with the different hotkeys. E.g.:</p>

<ul><li>Fuzzy:<ul><li>GlotPress: Ctrl-~ or Ctrl-Shift-~</li></ul><ul><li>GlotDict: Ctrl+Shift+F</li></ul></li></ul>

<p>We have some hotkeys with different actions. E.g.:</p>

<ul><li>Ctrl+Enter:<ul><li>GlotPress: Copy original.</li><li>GlotDict: Suggest or Save translation.</li></ul></li></ul>

<h2 id="options">Options</h2>

<p>The GlotPress hotkeys system is old (10 years) but it is not well document (I didn't know about it), so probably few people are using it. We have these options to improve the GlotPress and the <a href="https://[translate.wordpress.org](https://translate.wordpress.org/)/" target="_blank" rel="noopener">translate.wordpress.org</a> hotkeys:</p>

<ol><li>Add some hotkeys to GlotPress without taking care about the ones at GlotDict. </li><li>Integrate the GlotDict hotkeys in GlotPress. </li></ol>

<p>I think the best option in the second, but we have to take a decision about what we should do when we have a conflict with a hotkey between both systems: </p>

<ol><li>Maintain the GlotPress hotkeys.</li><li>Change it with the GlotDict hotkeys.</li><li>Set another one if it makes sense.</li></ol>

<p>In fact, it could make sense to rethink the whole hotkeys, asking the community members. E.g. some actions, like "Copy from original", could be rewritten with the Ctrl+Shift+C hotkey, similar to Ctrl+C (Windows/Linux) and Command+C (Mac). The current systems are using:</p>

<ul><li>GlotPress: Ctrl-Enter or Ctrl-Shift-B.</li><li>GlotDict: Ctrl+Shift+B.</li></ul>

<p>I think it would be useful to think about the usability in the main desktop operating systems (Windows, Linux, Mac) and in the desktop and laptop keyboards.</p>

<h2 id="improvements">Improvements</h2>

<p>It will be interesting to <a rel="noopener" href="https://github.com/GlotPress/GlotPress/issues/582" target="_blank">add a modal to show the keyboard shortcuts</a>, so the users will be informed about the hotkeys.</p>

**Comment by psmits1567** [comment]
Well I prefer to keep the list simple, but you have to bear in mind that people have other addons activated as well. Shortkeys from GlotDict and GlotPress and WPGPT and WP Translation Filler should not influence each other. Also there are a lot of other programs that could be open, so we need to be carefull not to define shortkeys in use by official programs like Excell, Word, etc.

**Reply by amieiro** [reply]
The hotkeys only work when you are in a page with GlotPress (it includes https://translate.wordpress.org/), not in all the system. This update will not conflict with other software installed on the system.

**Reply by vlad-timotei** [reply]
@psmits1567 Regarding a possible tools shortcuts conflict, WPGPT's shortcuts do not override default GlotPress ones and also respects the GlotDict ones. However, if the main project (GlotDict) introduces some sort of conflict, it's only fair for 3rd party tools to adjust - so if if it will be case I'd be happy to adjust any shortcut in WPGPT to maintain the default GlotPress ones.

**Comment by Mte90** [comment]
So as GlotDict author I am ok to migrate everything from GlotDict to GlotPress as the project scope was to prototype features in this way.

About the Modal I worked on the PR that was closed for that so I am ok to this too. We saw in GlotDict that people doesn't know about them so we added a panel inside the page to show what are the hotkeys, if we are working to change it in a better way, new hotkeys require the modal itself.

**Comment by vlad-timotei** [comment]
Love this idea a lot as my translation workflow heavily relies on shortcuts. There are 3 shortcuts we need to discuss. <br>
✅   means that GlotDict shortcut is already present in GlotPress shortcuts.




Action | Current GlotPress Hotkey | Issues | 
-- | -- | --
Cancel (Hide the editor) | `Escape` or `Ctrl + Shift + Z` | ✅
Previous editor | `Page Up` or `Ctrl + Up Arrow` | ✅
Next editor | `Page Down` or `Ctrl + Down Arrow` | ✅
| 1. Save | `Shift-Enter` |  GlotDict uses `Ctrl + Enter` (like Poedit does). <br> Options: <br> <br> 1. Add `Ctrl + Enter` as **alternative** GP shortcut. <br> In this case we would need to remove it as <br> shortcut for Copy original. (suggested option)<br><br> 2. No change in GlotPress. <br> Let GlotDict introduce the additional shortcut <br> for those used to it and keep the current one for the others. 
| 2. Copy original | `Ctrl-Enter` or `Ctrl + Shift + B` | Using `Ctrl + Shift + C` instead is very system alike. 👍🏻  <br><br> Regarding `Ctrl + Shift + B` that both GP and GD use  <br> we could either let it be or remove it. <br><br>
Approve | `Ctrl + +` or `Ctrl + Shift + A` | ✅
Reject | `Ctrl + -` or `Ctrl + Shift + R` | ✅
| 3. Fuzzy | `Ctrl + ~` or `Ctrl + Shift + ~` | `~` is a hard to obtain character on some keyboard layouts. <br>Maybe we need easier ones **instead**. <br><br> GlotDict introduces `Ctrl + Shift + F` <br> It is very alike with other actions <br>(Approve, Reject -> Fuzzy). <br> **(suggested shortcut)**<br><br>WPGPT introduces `Ctrl + *` <br>`*` is next to `+`, `-` on the numeric keyboard <br> Note: `Ctrl + +`, `Ctrl + -` only work on numeric keyboard.<br>_(optional alternative shortcut)_

**Comment by NekoJonez** [comment]
https://defkey.com/

This might be helpful in this thread

**Comment by amieiro** [comment]
<p>In this comment, I propose a solution to merge the hotkeys from both systems when it is possible.</p>



<p>When I talk about:</p>



<ul><li>"Ctrl", the Mac alternative will be "Command".</li><li>"Alt", the Mac alternative will be "Option".</li></ul>



<h2>Current GlotPress hotkeys</h2>



<figure class="wp-block-table"><table><tbody><tr><td><strong>Action</strong></td><td><strong>Proposed Hotkeys</strong></td><td><strong>Comments</strong></td><td><strong>Desktop<br>friendly</strong></td><td><strong>Laptop<br>friendly</strong></td></tr><tr><td>Cancel (Hide the editor)</td><td>- Escape<br>- Ctrl + Shift + Z</td><td>The 2nd option is used in both systems.</td><td>Yes</td><td>Yes</td></tr><tr><td>Previous string editor</td><td>- Page Up<br>- Ctrl + Up Arrow</td><td>The 1st option is used in both systems.</td><td>Yes</td><td>2nd option</td></tr><tr><td>Next string editor</td><td>- Page Down <br>- Ctrl + Down Arrow</td><td>The 1st option is used in both systems.</td><td>Yes</td><td>2nd option</td></tr><tr><td>Suggest or Save</td><td>- Ctrl + Enter</td><td>This command is used in GlotDict.<br>It is used in GlotPress as "Copy original" -&gt; I propose to delete this usage.</td><td>Yes</td><td>Yes</td></tr><tr><td>Copy from original</td><td>- Ctrl + Shift + B<br>- Ctrl + Shift + C</td><td>The 1st option is used in both systems.<br>2nd option is a new proposal.<br>I propose to remove the "Ctrl + Enter" hotkey for this usage. See previous row.</td><td>Yes</td><td>Yes</td></tr><tr><td>Approve</td><td>- Ctrl + + <br>- Ctrl + Shift + A</td><td>The 2nd option is used in both systems.</td><td>Yes</td><td>2nd option</td></tr><tr><td>Reject</td><td>- Ctrl + - <br>- Ctrl + Shift + R</td><td>The 2nd option is used in both systems.</td><td>Yes</td><td>2nd option</td></tr><tr><td>Fuzzy</td><td>- Ctrl + ~ <br>- Ctrl + Shift + ~<br>- Ctrl + Shift + F</td><td>I propose to add the 3rd option. The 1st and the 2nd are not usable in some systems. I only propose to maintain it by compatibility.</td><td>3rd option</td><td>3rd option</td></tr></tbody></table></figure>



<h2>New hotkeys</h2>



<p>This hotkeys are in use in GlotDict and make sense in GlotPress, so I propose to add it.</p>



<figure class="wp-block-table"><table><tbody><tr><td><strong>Action</strong></td><td><strong>Proposed Hotkeys</strong></td><td><strong>Comments</strong></td><td><strong>Desktop<br>friendly</strong></td><td><strong>Laptop<br>friendly</strong></td></tr><tr><td>Add Glossary definition in the translation area</td><td>- Right click on a Glossary term. </td><td>Add the first item from the glossary to the translation.</td><td>Yes</td><td>Yes</td></tr><tr><td>Dismiss validation warnings for the current visible editor</td><td>- Ctrl+D</td><td></td><td>Yes</td><td>Yes</td></tr><tr><td>Dismiss all validation warnings</td><td>- Ctrl + Shift + D</td><td></td><td>Yes</td><td>Yes</td></tr></tbody></table></figure>



<p>This hotkeys are very related with GlotDict, so I think it doesn't make sense to integrate now in GlotPress. May be, in a future, we can add some of these.</p>



<figure class="wp-block-table"><table><tbody><tr><td><strong>Action</strong></td><td><strong>Proposed Hotkeys</strong></td><td><strong>Comments</strong></td><td><strong>Desktop<br>friendly</strong></td><td><strong>Laptop<br>friendly</strong></td></tr><tr><td><strong>Add non-breaking spaces near symbols</strong></td><td>- Ctrl + Shift + X<br></td><td>This command is used mainly to comply with the <a rel="noreferrer noopener" href="https://fr.wordpress.org/team/handbook/traduire-wordpress-en-francais/les-regles-typographiques-utilisees-pour-la-traduction-de-wp-en-francais/" target="_blank">Typography rules used for WordPress translation</a> (French language).</td><td></td><td></td></tr><tr><td><strong>Force suggest or Force save translation</strong></td><td>- Ctrl+Shift+Enter</td><td>Dismiss the warnings and save. This hotkey is related with some GlotDict warnings, not with the GlotPress warnings.</td><td></td><td></td></tr><tr><td><strong>Load all consistency suggestions</strong></td><td>- Alt+C</td><td>This hotkey is related with the consistency suggestions (see image below), not present in GlotPress.</td><td></td><td></td></tr></tbody></table></figure>


<img width="2181" alt="Suggestions-from-Consistency" src="https://user-images.githubusercontent.com/1667814/165743897-3c423958-a334-4e9a-b7bc-9bdd59b0eb9e.png">

I propose to add a panel with information about the hotkeys. Something like you can see in the next screenshot, but probably with a lightbox:

![image](https://user-images.githubusercontent.com/1667814/165744164-7b7da1a5-b476-439f-92cd-97beece6b274.png)

**Comment by amieiro** [comment]
The https://github.com/GlotPress/GlotPress/pull/1456/ PR implements what we discuss here.
